### ISSUE TITLE
nalgebra and differential-dataflow regressed

### ISSUE THREAD

**Issue by Zoxc** [rust-lang/rust#42003]
nalgebra v0.10.1 and differential-dataflow v0.0.3 works on 5aaa60611cb7d89b03da8e56653ce09b0203f81d, but are broken on 9f15631c36665911eb8e6f594ebcfe93e65a461c.

UPDATE: nalgebra fixed, different-dataflow not yet.

**Comment by est31** [comment]
broken in which way? do you have an error message or ICE you can share with us?

**Comment by Zoxc** [comment]
nalgebra:
```
   Compiling nalgebra v0.10.1
error[E0391]: unsupported cyclic reference between types/traits detected
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\nalgebra-0.10.1\src\structs\dmatrix.rs:148:5
    |
148 |     mij:   [N; 2 * 2],
    |     ^^^^^^^^^^^^^^^^^ cyclic reference
    |
note: the cycle begins when processing `structs::dmatrix::DMatrix2::mij`...
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\nalgebra-0.10.1\src\structs\dmatrix.rs:148:5
    |
148 |     mij:   [N; 2 * 2],
    |     ^^^^^^^^^^^^^^^^^
note: ...which then requires const-evaluating `structs::dmatrix::DMatrix2::{{initializer}}`...
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\nalgebra-0.10.1\src\structs\dmatrix.rs:148:16
    |
148 |     mij:   [N; 2 * 2],
    |                ^^^^^
note: ...which then requires processing `structs::dmatrix::DMatrix2::{{initializer}}`...
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\nalgebra-0.10.1\src\structs\dmatrix.rs:148:16
    |
148 |     mij:   [N; 2 * 2],
    |                ^^^^^
note: ...which then requires coherence checking all impls of trait `std::ops::Mul`...
note: ...which then requires processing `structs::dmatrix::DMatrix`...
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\nalgebra-0.10.1\src\structs\dmatrix.rs:20:1
    |
20  | / pub struct DMatrix<N> {
21  | |     nrows: usize,
22  | |     ncols: usize,
23  | |     mij:   Vec<N>
24  | | }
    | |_^
note: ...which then requires computing the variances for items in this crate...
    = note: ...which then again requires processing `structs::dmatrix::DMatrix2::mij`, completing the cycle.
```
differential-dataflow:
```
   Compiling differential-dataflow v0.0.3
error: no associated item named `new` found for type `timely_sort::Sorter<_>` in the current scope
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\differential-dataflow-0.0.3\src\operators\arrange.rs:122:26
    |
122 |         let mut sorter = LSBRadixSorter::new();
    |                          ^^^^^^^^^^^^^^^^^^^
    |
    = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
    = help: candidate #1: `use timely_sort::RadixSorterBase;`

error: no associated item named `new` found for type `timely_sort::Sorter<_>` in the current scope
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\differential-dataflow-0.0.3\src\operators\arrange.rs:222:26
    |
222 |         let mut sorter = LSBRadixSorter::new();
    |                          ^^^^^^^^^^^^^^^^^^^
    |
    = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
    = help: candidate #1: `use timely_sort::RadixSorterBase;`

error: no associated item named `new` found for type `timely_sort::Sorter<_>` in the current scope
   --> C:\Users\John\.cargo\registry\src\github.com-1ecc6299db9ec823\differential-dataflow-0.0.3\src\operators\consolidate.rs:110:70
    |
110 |                 let entry = inputs.entry_or_insert(index.time(), || (LSBRadixSorter::new(), 0, default_threshold));
    |                                                                      ^^^^^^^^^^^^^^^^^^^
    |
    = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
    = help: candidate #1: `use timely_sort::RadixSorterBase;`
```

**Comment by TimNN** [comment]
Looks like a duplicate of #41936, hopefully fixed by #41937 once it lands.

**Comment by frewsxcv** [comment]
Is this fixed now?

**Comment by Zoxc** [comment]
I think `nalgebra` is fixed now. It may not compile with `RUST_LOG=rustc_typeck::check::regionck` due to that failing with `rustc_serialize`.

**Comment by nikomatsakis** [comment]
What about  differential-dataflow ?

Note that compiling with `RUST_LOG` is not really a *supported* workflow, though the current setup is generating a lot of cycle errors in those cases and needs to be fixed.

**Comment by Zoxc** [comment]
`nalgebra` compiles on `rustc 1.19.0-nightly (d015610db 2017-06-05)`, `differential-dataflow` gives the same error.

**Comment by nikomatsakis** [comment]
OK. Regression. Nominating.

**Comment by nikomatsakis** [comment]
cc @rust-lang/compiler -- regression, anybody have any ideas?

**Comment by eddyb** [comment]
`differential-dataflow` is a completely different error, *and a much more serious one*. Either traits-in-scope is broken or there's an inherent method that's not found.

**Comment by est31** [comment]
I've tried to bisect the regression, but all I could find was rust compiler versions where the compilation failed.

* rust 1.11 stable fails compilation
* rust 1.16 stable fails compilation
* rust 1.17 stable fails compilation
* commit 4465f22ef6498db65059a41595d354a7af56cbea fails compilation, including all commits down to 927c55d86b0be44337f37cf5b0a76fb8ba86e06c where uploading on each merge started.

Therefore, @Zoxc can you please point to a rustc version where compilation works?

**Comment by Zoxc** [comment]
It did work on a crater run on top of https://github.com/rust-lang/rust/commit/5aaa60611cb7d89b03da8e56653ce09b0203f81d. That looks pretty close to 1.16 stable though.

**Comment by Zoxc** [comment]
Version `0.0.3` was published on `crates.io` Jun 7, 2016, so it should probably work on rust 1.9.

**Comment by est31** [comment]
Still fails compilation on 1.9: http://paste.ubuntu.com/24804993/
And on 1.8 as well.

**Comment by Zoxc** [comment]
I built https://github.com/rust-lang/rust/commit/5aaa60611cb7d89b03da8e56653ce09b0203f81d locally and it doesn't work there either :/

**Comment by nikomatsakis** [comment]
A bit surprising that we've not seen this in cargobomb runs or other things. I'm inclined to believe that was a bug in that release of differential-dataflow -- but hard to say without closer investigation
.
cc @frankmcsherry -- any thoughts? we're looking at a (potential, unverified) regression in v0.0.3 of differential-dataflow (I see you're up to 0.1.0 now)

**Comment by nikomatsakis** [comment]
triage: P-high

I'm marking this P-high, but mostly because I want to try and confirm if there has been a regression here or not (hence no assignee)

**Comment by frankmcsherry** [comment]
When `PartialOrd` was ret-conned to mean "the same as `Ord`, if `Ord` exists" all correctness for 0.0.3 went out the window. The behavior of sorting changed for types where `partial_cmp` was not identical to `cmp`, and I'm sure that there was fall-out.

**Comment by nikomatsakis** [comment]
@frankmcsherry hmm, ok. This is a compilation error though.

**Comment by frankmcsherry** [comment]
Ah, sorry. The reason I haven't been tracking it is for that reason. The reason there is a compile error above is because the timely_sort crate had a breaking change. It got fixed in master, but can't fix 0.0.3.

**Comment by nikomatsakis** [comment]
@frankmcsherry thanks! :) I'm going to close this bug then.
